### PR TITLE
#13319 final productprice

### PR DESCRIPTION
**Pull request by Chemssou**
https://github.com/metasfresh/metasfresh/issues/13339

### CODE REVIEW

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+import java.io.File;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.List;
+
+public class CostRevaluationDocumentHandler implements DocumentHandler
+{
+
+	private static I_M_CostRevaluation extractForecast(final DocumentTableFields docFields)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java:43]: rename `extractForecast` to `extractRecord`

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+package de.metas.costrevaluation;
+
+import de.metas.util.ISingletonService;
+import org.compiere.model.I_M_CostRevaluationLine;
+
+import java.util.List;
+
+public interface ICostRevaluationDAO  extends ISingletonService
+{
+	List<I_M_CostRevaluationLine> retrieveLinesByCostRevaluationId(int costRevaluationId);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java:32]: pls introduce CostRevaluationId (implements RepoIdAware).

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+	I_M_CostRevaluationLine getCostRevaluationLineById (int costRevaluationLineId);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java:34]: pls introduce CostRevaluationLineId similar with BPartnerLocationId,
i.e. basically will contain the CostRevaluationId and repoId (of the line)

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java (hunk 2 above)
Comment (teosarca) [reply backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java:32]: just saw that u already introduced it. pls use it.

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java (hunk 3 above)
Comment (teosarca) [reply backend/de.metas.business/src/main/java/de/metas/costrevaluation/ICostRevaluationDAO.java:34]: just saw that u already introduced it. pls use it.

File: backend/de.metas.business/src/main/sql/postgresql/system/20-de.metas.business/5647850_sys_gh13319_final_product_price.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+ALTER TABLE m_costrevaluation DROP COLUMN name;
+ALTER TABLE m_costrevaluation DROP COLUMN description;
+
+ALTER TABLE m_costrevaluationline DROP COLUMN name;
+ALTER TABLE m_costrevaluationline DROP COLUMN description;
+ALTER TABLE m_costrevaluationline DROP COLUMN cm_template_id;
+ALTER TABLE m_costrevaluationline DROP COLUMN otherclause;
+ALTER TABLE m_costrevaluationline DROP COLUMN whereclause;
+ALTER TABLE m_costrevaluationline DROP COLUMN ad_table_id;
```
Comment (teosarca) [inline backend/de.metas.business/src/main/sql/postgresql/system/20-de.metas.business/5647850_sys_gh13319_final_product_price.sql:30]: FYI when u create a new table you can copy one of our template tables to speed things up
| tablename |
| :--- |
| X\_AcctDocTableTemplate |
| X\_EventStoreTemplate |
| X\_ImportTableTemplate |
| X\_TableTemplate |


most probably you will want `X_TableTemplate`.

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	@Override
+	public String completeIt(final DocumentTableFields docFields)
+	{
+		final I_M_CostRevaluation costRevaluation = extractForecast(docFields);
+
+		final int costRevaluationId = costRevaluation.getM_CostRevaluation_ID();
+		final List<I_M_CostRevaluationLine> lines = costRevaluationDAO.retrieveLinesByCostRevaluationId(costRevaluationId);
+		if (lines.size() <= 0)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java:82]: use isEmpty()

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		costRevaluation.setDocAction(IDocument.ACTION_None);
+		return IDocument.STATUS_Completed;
+	}
+
+	@Override
+	public void reactivateIt(final DocumentTableFields docFields)
+	{
+		final I_M_CostRevaluation costRevaluation = extractForecast(docFields);
+		costRevaluation.setProcessed(false);
+		costRevaluation.setDocAction(IDocument.ACTION_Complete);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java:96]: throw new UnsupportedOperationException();

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/impl/CostRevaluationLineId.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import de.metas.util.Check;
+import de.metas.util.lang.RepoIdAware;
+import lombok.NonNull;
+import lombok.Value;
+
+import javax.annotation.Nullable;
+
+@Value
+public class CostRevaluationLineId implements RepoIdAware
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/impl/CostRevaluationLineId.java:33]: check if the de.metas.util.lang.All_RepoIdAware_Classes_Test test is failing.
I think you might need to skip `CostRevaluationLineId`.

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments, check why jenkins fails

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+public class CostRevaluationDocumentHandler implements DocumentHandler
+{
+
+	private static I_M_CostRevaluation extractForecast(final DocumentTableFields docFields)
+	{
+		return InterfaceWrapperHelper.create(docFields, I_M_CostRevaluation.class);
+	}
+
+	private final ICostRevaluationDAO costRevaluationDAO = Services.get(ICostRevaluationDAO.class);
```
Comment (metas-rc) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationDocumentHandler.java:48]: Please, declare the services at the beginning of the class, before adding the methods.

**Review by metas-rc — COMMENTED** [review]
How did you solve the document number automatic setting?

File: backend/de.metas.business/src/main/java/de/metas/costrevaluation/impl/CostRevaluationDAO.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+		return queryBL.createQueryBuilder(I_M_CostRevaluationLine.class)
+				.addEqualsFilter(I_M_CostRevaluationLine.COLUMN_M_CostRevaluation_ID, costRevaluationId)
+				.filter(ActiveRecordQueryFilter.getInstance(I_M_CostRevaluationLine.class))
+				.orderBy(I_M_CostRevaluationLine.COLUMN_M_CostRevaluationLine_ID)
+				.create()
+				.list();
+	}
+
+	@Override
+	public I_M_CostRevaluationLine getCostRevaluationLineById(final int costRevaluationLineId)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/costrevaluation/impl/CostRevaluationDAO.java:55]: where is this method used?
if it's not used, i suggest to delete it.

File: backend/de.metas.business/src/main/sql/postgresql/system/20-de.metas.business/5647850_sys_gh13319_final_product_price.sql (hunk 4 above)
Comment (Chemssou) [reply backend/de.metas.business/src/main/sql/postgresql/system/20-de.metas.business/5647850_sys_gh13319_final_product_price.sql:30]: I used the `TemplateTable`  without the X, i'll consider that in the future. thanks

**Comment by Chemssou** [comment]
> How did you solve the document number automatic setting?

I used a  a `DefaultValue` =`< >`   in the `documentNo` column , cause after save the proper sequence is added.

**Comment by Chemssou** [comment]
> check my comments, check why jenkins fails

- Comments addressed one by one.
- The build fails,  cause branch name contains a special char `#` as i created a new one without `#` and build OK.

**Comment by metas-rc** [comment]
> > How did you solve the document number automatic setting?
> 
> I used a `DefaultValue` =`< >` in the `documentNo` column , cause after save the proper sequence is added. sort of a workaround.

Couldn't this de.metas.document.sequence.impl.DocumentNoBuilder be used? I would like to have it properly done.

**Comment by metas-rc** [comment]
> [metasfresh/is184#16](https://github.com/metasfresh/is184/issues/16)

Please, don't put customer links in the metasfresh issues and/or pull requests ❗

**Comment by metas-rc** [comment]
> > check my comments, check why jenkins fails
> 
> * Comments addressed one by one.
> * The build fails,  cause branch name contains a special char `#` as i created a new one without `#` and build OK.

where is it?

**Comment by Chemssou** [comment]
> where is it? Please, put it in the me03 issue.
image

**Comment by metas-rc** [comment]
> > where is it? Please, put it in the me03 issue.
> > image

@Chemssou please, don't hide it. Strike the first one with a line and put the new one exactly under it. I want to see the correct PR as soon as I check the R&I  request.

**Pull request closed without merging**